server/start.go

```diff
@@ -590,15 +597,71 @@ func createAndStartIndexerService(
 }
 
 // getAppGenesis returns a function which returns the app genesis based on its location in the config.
-func getAppGenesis(cfg *cmtcfg.Config) func() (*genutiltypes.AppGenesis, error) {
-	return func() (*genutiltypes.AppGenesis, error) {
-		appGenesis, err := genutiltypes.AppGenesisFromFile(cfg.GenesisFile())
+func getAppGenesis(cfg *cmtcfg.Config) (*genutiltypes.AppGenesis, uint64, error) {
+	genFile, err := os.Open(filepath.Clean(cfg.GenesisFile()))
+	if err != nil {
+		return nil, 0, err
+	}
+	defer genFile.Close()
+
+	reader := bufio.NewReader(genFile)
+
+	appGenesis, err := genutiltypes.AppGenesisFromReader(reader)
+	if err != nil {
+		return nil, 0, fmt.Errorf("failed to read genesis from file %s: %w", genFile.Name(), err)
+	}
+
+	daStartHeight, err := parseDAStartHeightFromGenesis(reader)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return appGenesis, daStartHeight, nil
+}
+
+const DAStartHeightFieldName = "da_start_height"
+
+// parseDAStartHeightFromGenesis parses the `da_start_height` from a genesis JSON file, aborting early after finding the `da_start_height`.
+// For efficiency, it's recommended to place the `da_start_height` field before any large entries in the JSON file.
+// Returns no error when the `da_start_height` field is not found.
+// Logic based on https://github.com/cosmos/cosmos-sdk/blob/v0.50.14/x/genutil/types/chain_id.go#L18.
+func parseDAStartHeightFromGenesis(r io.Reader) (uint64, error) {
```

Comment:
the manual parsing seems excessive here,  `parseDAStartHeightFromGenesis` is called from `getAppGenesis` once before the node starts right? Can't we just unmarshal it into a map and access the field?

Reply:
I see, makes sense. I assume it is still potentially expensive to unmarshal a large genesis file into a struct that just has a single field.